Forge Networking Remastered Unity/Assets/Bearded Man Studios Inc/Scripts/Networking/ArgumentParser.cs

```diff
@@ -1,31 +1,84 @@
-﻿using System.Collections.Generic;
+﻿using System;
+using System.Collections.Generic;
+using System.Text.RegularExpressions;
 
 namespace BeardedManStudios
 {
-	// A simple dummy class for parsing the string arguments of a generic program, including flags
+	/// <summary>
+	/// A simple dummy class for parsing the string arguments of a generic program, including flags
+	/// </summary>
 	public static class ArgumentParser
 	{
 		/// <summary>
-		/// Parse a string array for the various arguments that were supplied
+		/// Matches:
+		///		-key
+		///		-key=value
+		///		--key
+		///		--key=value
+		/// </summary>
+		private const string ARGUMENT_PATTERN = @"^-{1,2}(?'key'[^\s=]+)(=(?'value'[^\s-=]+))?$";
+
+		/// <summary>
+		/// Parse a string array for the various arguments that were supplied.
+		/// For example:
+		///		["--a=1", "-b=2", "--c", "-d", "e", "f", "g"]
+		/// Results in:
+		///		{"a":"1", "b":"2", "c":null, "d":null, "0":"e", "1":"f", "2":"g"}
 		/// </summary>
 		/// <param name="args">The arguments that were sent to the program</param>
-		/// <returns>A dictionary of arguments with a key being the index or the flag starting with "-"</returns>
-		public static Dictionary<string, string> Parse(string[] args)
+		/// <returns>
+		/// A dictionary of arguments with the key being the flag (or the index for floating args)
+		/// and the value being the associated value (or argument for floating args).
+		/// </returns>
+		public static Dictionary<string, string> Parse(params string[] args)
 		{
 			Dictionary<string, string> parsedArgs = new Dictionary<string, string>();
 
 			int floatingArgIndex = 0;
 			for (int i = 0; i < args.Length; i++)
 			{
-				if (args[i].StartsWith("--"))
-					parsedArgs.Add(args[i], args[i]);
-				else if (args[i].StartsWith("-"))
-					parsedArgs.Add(args[i++], args[i]);
-				else
-					parsedArgs.Add((floatingArgIndex++).ToString(), args[i]);
+				string arg = args[i];
+
+				string value;
+				string key = ParseArg(arg, out value) ?? floatingArgIndex++.ToString();
+
+				parsedArgs[key] = value;
 			}
 
 			return parsedArgs;
 		}
+
+		/// <summary>
+		/// Given an input argument in one of the following formats:
+		/// 	-key
+		///		-key=value
+		///		--key
+		///		--key=value
+		///		value
+		/// 
+		/// Produces the following key/value pair:
+		///		{"key", null}
+		///		{"key", "value"}
+		///		{"key", null}
+		///		{"key", "value"}
+		///		{null, "value"}
+		/// </summary>
+		/// <param name="arg"></param>
+		/// <param name="value"></param>
+		/// <returns></returns>
+		private static string ParseArg(string arg, out string value)
```

Comment:
Maybe rename this method to `ParseArguments` :thinking: ?

Reply:
Done